Nodejs/Product/Nodejs/Workspace/JsFileDebugLaunchTargerProvider.cs

```diff
@@ -0,0 +1,102 @@
+            }
+        }
+
+        private void SetupDebugTargetInfoForNodeProtocol(ref VsDebugTargetInfo vsDebugTargetInfo, DebugLaunchActionContext debugLaunchContext, string nodeExe)
+        {
+            var target = $"\"{vsDebugTargetInfo.bstrExe}\"";
+            if (debugLaunchContext.LaunchConfiguration.TryGetValue<string>(ScriptArgsKey, out var scriptArgs) && !string.IsNullOrEmpty(scriptArgs))
+            {
+                target += " " + scriptArgs;
+            }
+            vsDebugTargetInfo.bstrArg = target;
+
+            vsDebugTargetInfo.bstrExe = nodeExe;
+            if (debugLaunchContext.LaunchConfiguration.TryGetValue<string>(NodeArgsKey, out var nodeArgs) && !string.IsNullOrWhiteSpace(nodeArgs))
+            {
+                AppendOption(ref vsDebugTargetInfo, AD7Engine.InterpreterOptions, nodeArgs);
+            }
+
+            if (debugLaunchContext.LaunchConfiguration.TryGetValue<string>(DebuggerPortKey, out var debuggerPort) && !string.IsNullOrEmpty(debuggerPort))
+            {
+                AppendOption(ref vsDebugTargetInfo, AD7Engine.DebuggerPort, debuggerPort);
+            }
+            AppendOption(ref vsDebugTargetInfo, AD7Engine.WaitOnNormalExitSetting, "true"); // todo: make this an option
+
+            vsDebugTargetInfo.clsidCustom = AD7Engine.DebugEngineGuid;
+            vsDebugTargetInfo.grfLaunch = (uint)__VSDBGLAUNCHFLAGS.DBGLAUNCH_StopDebuggingOnEnd;
+
+            void AppendOption(ref VsDebugTargetInfo dbgInfo, string option, string value)
+            {
+                if (!string.IsNullOrWhiteSpace(dbgInfo.bstrOptions))
+                {
+                    dbgInfo.bstrOptions += ";";
+                }
+
+                dbgInfo.bstrOptions += option + "=" + HttpUtility.UrlEncode(value);
+            }
+        }
+
+        private void SetupDebugTargetInfoForInspectProtocol(ref VsDebugTargetInfo vsDebugTargetInfo, DebugLaunchActionContext debugLaunchContext, string nodeExe)
+        {
+            var target = vsDebugTargetInfo.bstrExe;
+            var cwd = vsDebugTargetInfo.bstrCurDir;
+
+            var jsonContent = GetJsonConfigurationForInspectProtocol(target, cwd, nodeExe, debugLaunchContext);
+
+            vsDebugTargetInfo.dlo = DEBUG_LAUNCH_OPERATION.DLO_CreateProcess;
+            vsDebugTargetInfo.clsidCustom = NodejsProjectLauncher.WebKitDebuggerV2Guid;
+            vsDebugTargetInfo.bstrExe = target;
+            vsDebugTargetInfo.bstrOptions = jsonContent;
+            vsDebugTargetInfo.grfLaunch = (uint)__VSDBGLAUNCHFLAGS.DBGLAUNCH_StopDebuggingOnEnd;
+        }
+
+        bool ILaunchDebugTargetProvider.SupportsContext(IWorkspace workspaceContext, string targetFilePath) => true;
+
+        void ILaunchDebugTargetProvider.LaunchDebugTarget(IWorkspace workspaceContext, IServiceProvider serviceProvider, DebugLaunchActionContext debugLaunchActionContext)
+        {
+            // nope
```

Comment:
Add assertion?

Reply:
It's called we just don't have any work to do, I'll update the comment